research/gnn-survey/layers.py

```diff
@@ -183,3 +185,74 @@ def call(self, inputs):
 
     output = tf.sparse.sparse_dense_matmul(attn, x)
     return output
+
+
+class GraphIsomorphismLayer(tf.keras.layers.Layer):
+  """Single graph isomorphism layer."""
+
+  def __init__(self, mlp_layers, output_dim, dropout_rate, learn_eps=False, sparse=True, **kwargs):
+    """Initializes the GraphIsomorphismLayer.
+
+    Args:
+      mlp layers: (int) Number of mlp layers
+      output_dim: (int) Output dimension of gcn layer
+      dropout_rate: (float) Dropout probability
+      learn_eps: (bool) Whether to learn the epsilon weighting
+      sparse: (bool) sparse: Whether features are sparse
+      **kwargs: Keyword arguments for tf.keras.layers.Layer.
+    """
+    super(GraphIsomorphismLayer, self).__init__(**kwargs)
+    self.mlp_layers = mlp_layers
+    self.output_dim = output_dim
+    self.dropout_rate = dropout_rate
+    self.learn_eps = learn_eps
+    self.sparse = sparse
+
+    self.relu = tf.keras.layers.ReLU()
+    self.dropout = tf.keras.layers.Dropout(self.dropout_rate)
+    self.batch_norm = []
+    for i in range(self.mlp_layers-1):
+      self.batch_norm.append(tf.keras.layers.BatchNormalization())
+
+  def build(self, input_shape):
+    super(GraphIsomorphismLayer, self).build(input_shape)
+    self.weight = []
+    for i in range(self.mlp_layers):
+      if i == 0:
+        self.weight.append(
+          self.add_weight(
+              name='weight'+str(i+1),
+              shape=(input_shape[0][-1], self.output_dim),
+              initializer='random_normal',
+              trainable=True)
+          )
+      else:
+        self.weight.append(
+          self.add_weight(
+              name='weight'+str(i+1),
+              shape=(self.output_dim, self.output_dim),
+              initializer='random_normal',
```

Comment:
It's not clear that you actually want random_normal here. glorot_uniform or glorot_normal may be preferable since they scale by the shape dimensions, which is why it's the Keras default: https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense

Reply:
OK! Should I replace all weight to glorot_normoal initializer? or just GIN layer need to replace? 
Thanks:)

Reply:
If you can replace all of them, that would be great.

Reply:
OK:)